crates/git_ui/src/clone.rs

```diff
@@ -0,0 +1,207 @@
+    Render, SharedString, Window, div,
+};
+use std::sync::Arc;
+use ui::{Color, CommonAnimationExt, Icon, IconName, Label, h_flex, prelude::*, v_flex};
+use util::ResultExt;
+use workspace::{self, AppState, ModalView};
+
+struct CloneProgressModal {
+    repo_name: SharedString,
+    focus_handle: FocusHandle,
+}
+
+impl CloneProgressModal {
+    fn new(repo_name: String, cx: &mut Context<Self>) -> Self {
+        Self {
+            repo_name: repo_name.into(),
+            focus_handle: cx.focus_handle(),
+        }
+    }
+}
+
+impl Focusable for CloneProgressModal {
+    fn focus_handle(&self, _: &gpui::App) -> FocusHandle {
+        self.focus_handle.clone()
+    }
+}
+
+impl Render for CloneProgressModal {
+    fn render(&mut self, _window: &mut Window, cx: &mut Context<Self>) -> impl IntoElement {
+        div().elevation_3(cx).w(ui::rems(34.)).p_4().child(
+            v_flex()
+                .gap_3()
+                .child(
+                    h_flex()
+                        .gap_2()
+                        .items_center()
+                        .child(
+                            Icon::new(IconName::ArrowCircle)
+                                .size(ui::IconSize::Small)
+                                .color(Color::Accent)
+                                .with_rotate_animation(2),
+                        )
+                        .child(
+                            Label::new(format!("Cloning \"{}\"", self.repo_name))
+                                .size(ui::LabelSize::Large),
+                        ),
+                )
+                .child(
+                    Label::new("Please wait while the repository is being cloned...")
+                        .size(ui::LabelSize::Small)
+                        .color(Color::Muted),
+                ),
+        )
+    }
+}
+
+impl EventEmitter<DismissEvent> for CloneProgressModal {}
+impl ModalView for CloneProgressModal {}
+
+pub fn clone_and_open(
+    repo_url: String,
+    app_state: Arc<AppState>,
+    cx: &mut App,
+    on_success: Arc<
+        dyn Fn(&mut workspace::Workspace, &mut Window, &mut Context<workspace::Workspace>)
+            + Send
+            + Sync
+            + 'static,
+    >,
+) {
+    workspace::with_active_or_new_workspace(cx, |_workspace, window, cx| {
+        let path_prompt = cx.prompt_for_paths(gpui::PathPromptOptions {
+            files: false,
+            directories: true,
+            multiple: false,
+            prompt: Some("Select directory for cloned repository".into()),
+        });
+
+        let workspace_weak = cx.weak_entity();
+        cx.spawn_in(window, async move |workspace, cx| {
+            let result = async {
+                let mut paths = path_prompt.await.ok()?.ok()??;
+                let path = paths.pop()?;
+
+                let repo_name = repo_url
+                    .split('/')
+                    .next_back()
+                    .and_then(|name| name.strip_suffix(".git"))
```

Comment:
I think we should probably have this support URLs that don't end with `.git`, so maybe `.next_back().map(|repo_name| repo_name.trim_end_matches(".git")).unwrap_or("repository")`